### PR TITLE
oop-object_independence-medium-q5

### PR DESCRIPTION
**Pull request by KaiKhademi**
## PR Checklist
- [x]  Are helper methods really needed here?
    > Opt for hardcoding input/output into the input provider when possible
- [x]  Are the assertion fail messages good?
    - [x]  Logical (does it make sense?)
    - [x]  Readable (can a student understand what’s wrong?)
    - [x]  Not too feed-y (does it give away the answer?)
    - [x]  Grammatically sound (is the grammar and punctuation in the sentence proper?)
    - [x]  Specific (does the fail message relate to the question?)
- [x]  Is the code formatted well?
    > Run `CTRL` + `ALT` + `L` in IntelliJ
- [x]  EOF newlines
- [x]  Do the comments for Parsons follow the correct convention?
- [x]  Is the question being tested sufficiently?
    - [x]  Are there representative cases to test the program?
    - [x]  What about edge cases? (empty strings, negatives, infinity, off-by-one errors, etc)
        > If the question doesn’t specify what to do, mark it as a bad question
- [x]  Are the test method names good?
    > Test method names show up in a “What went well” section on the website, so “mathTest” will become “What went well: math test”, which doesn’t make much sense. A better name in this example would be “calculatesTaxCorrectly” so it shows up as “What went well: calculates tax correctly”

### CODE REVIEW

File: src/oop/object_independence/medium/q5/QuestionText.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+Your `Team` class must have the following items:
+
+- Attributes:
+    - A `String` called **name**
+    - An `int` called **totalWins**
+    - Note that neither of these attributes should be visible from another class
+
+- Constructors:
+    - A single constructor with parameters of a `String` and an `itn` that initialises the **name** and
```
Comment (SamamthaH) [inline src/oop/object_independence/medium/q5/QuestionText.md:23]:
```suggestion
    - A single constructor with parameters of a `String` and an `int` that initializes the **name** and
```

File: src/oop/object_independence/medium/q5/QuestionText.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+      **totalWins** attributes
+    - This constructor should be visible from another class
+
+- Methods:
+    - a `getTotalWins` method which is accessible from other classes and returns the **totalWins** attribute
+    - a `setTotalWins` method which accepts an `int` value, updates the **totalWins** attribute with it,
+      and is accessible from other classes
+    - a `toString` method which returns a String of the information about the `Team` class of the form
+      `"Team " + name + " has won " + totalWins + " games"` and is visible from other classes.
```
Comment (SamamthaH) [inline src/oop/object_independence/medium/q5/QuestionText.md:32]: change this to be the x y formatting for consistency

File: src/oop/object_independence/medium/q5/MainTest.java [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+import org.junit.jupiter.params.provider.MethodSource;
+
+import java.util.stream.Stream;
+
+import static global.tools.CustomAssertions._assertArrayEquals;
+import static global.tools.CustomAssertions._assertEquals;
+import static org.junit.jupiter.api.Assertions.*;
+
+public class MainTest {
+    // Parsons question
```
Comment (SamamthaH) [inline src/oop/object_independence/medium/q5/MainTest.java:20]: incorrect formatting

File: src/oop/object_independence/medium/q5/MainTest.java [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    private final String className = "Team";
+    private ObjectTest testClass;
+    private final String firstFieldName = "name";
+    private final String secondFieldName = "totalWins";
+    private final String getSecondField = "getTotalWins";
+    private final String setSecondField = "setTotalWins";
+
+    @BeforeEach
+    public void setUp() {
+        String bankString = "oop.object_independence.medium.q5." + className;
```
Comment (SamamthaH) [inline src/oop/object_independence/medium/q5/MainTest.java:30]: incorrect name

File: src/oop/object_independence/medium/q5/MainTest.java [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+        testClass = new ObjectTest(bankString);
+    }
+
+    @Test
+    public void teamClassHasRequiredFields() {
+        assertTrue(testClass.hasField(firstFieldName, String.class),
+                "Your " + className + " class is missing the " + firstFieldName + " attributes.");
```
Comment (SamamthaH) [inline src/oop/object_independence/medium/q5/MainTest.java:37]:
```suggestion
                "Your " + className + " class is missing the " + firstFieldName + " attribute.");
```

File: src/oop/object_independence/medium/q5/MainTest.java [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+        assertTrue(testClass.hasModifier(firstFieldName, "private"),
+                "Your " + firstFieldName + " does not have the correct visibility modifier.");
```
Comment (SamamthaH) [inline src/oop/object_independence/medium/q5/MainTest.java:39]:
```suggestion
                "Your " + firstFieldName + " attribute does not have the correct visibility modifier.");
```

File: src/oop/object_independence/medium/q5/MainTest.java [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+        assertTrue(testClass.hasField(secondFieldName, int.class),
+                "Your " + className + " class is missing the " + secondFieldName + " attributes.");
```
Comment (SamamthaH) [inline src/oop/object_independence/medium/q5/MainTest.java:41]:
```suggestion
                "Your " + className + " class is missing the " + secondFieldName + " attribute.");
```

File: src/oop/object_independence/medium/q5/MainTest.java [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+        assertTrue(testClass.hasModifier(secondFieldName, "private"),
+                "Your " + secondFieldName + " does not have the correct visibility modifier.");
```
Comment (SamamthaH) [inline src/oop/object_independence/medium/q5/MainTest.java:43]:
```suggestion
                "Your " + secondFieldName + " attribute does not have the correct visibility modifier.");
```

File: src/oop/object_independence/medium/q5/MainTest.java [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+
+    @ParameterizedTest
+    @MethodSource("inputProvider")
+    public void teamClassHasCorrectConstructor(String value1, int value2) throws Throwable {
+        Object[][] arguments = {
+                {value1, String.class},
+                {value2, int.class}
+        };
+        Class<?>[] constructorClasses = {String.class, int.class};
+        assertTrue(testClass.hasConstructor(constructorClasses, new String[]{"public"}),
```
Comment (SamamthaH) [inline src/oop/object_independence/medium/q5/MainTest.java:72]: this already gets checked above

File: src/oop/object_independence/medium/q5/MainTest.java [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+                "Your " + className + " class is missing a required constructor.");
+        Object classInstance = testClass.createInstance(arguments);
+        _assertEquals(value1, testClass.getFieldValue(classInstance, firstFieldName),
+                "Your " + className + " constructor does not correctly initialize the " + firstFieldName + " field.");
```
Comment (SamamthaH) [inline src/oop/object_independence/medium/q5/MainTest.java:76]:
```suggestion
                "Your " + className + " constructor does not correctly initialize the " + firstFieldName + " attribute.");
```

File: src/oop/object_independence/medium/q5/MainTest.java [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+        _assertEquals(value2, testClass.getFieldValue(classInstance, secondFieldName),
+                "Your " + className + " constructor does not correctly initialize the " + secondFieldName + " field.");
```
Comment (SamamthaH) [inline src/oop/object_independence/medium/q5/MainTest.java:78]:
```suggestion
                "Your " + className + " constructor does not correctly initialize the " + secondFieldName + " attribute.");
```

File: src/oop/object_independence/medium/q5/MainTest.java [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+    @MethodSource("inputProvider")
+    public void teamClassHasCorrectGetTotalWinsMethod(String value1, int value2) throws Throwable {
+        Object[][] arguments = {
+                {value1, String.class},
+                {value2, int.class}
+        };
+        Object classInstance = testClass.createInstance(arguments);
+        String[] getMethodModifiers = {"public"};
+        String incorrectGetMethodMessage = String.join(" ",
+                "Your", getSecondField, "does not correctly get the value of the", secondFieldName, "attribute.");
```
Comment (SamamthaH) [inline src/oop/object_independence/medium/q5/MainTest.java:91]:
```suggestion
                "Your", getSecondField, "method does not correctly get the value of the", secondFieldName, "attribute.");
```

File: src/oop/object_independence/medium/q5/MainTest.java [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+                {"Name", String.class},
+                {value, int.class}
+        };
+        Object classInstance = testClass.createInstance(arguments);
+        Object[][] setArguments = {
+                {update, int.class}
+        };
+        String[] setMethodModifiers = {"public"};
+        String incorrectSetMethodMessage = String.join(" ",
+                "Your", setSecondField, "does not correctly set the value of the", secondFieldName, "attribute.");
```
Comment (SamamthaH) [inline src/oop/object_independence/medium/q5/MainTest.java:118]:
```suggestion
                "Your", setSecondField, "method does not correctly set the value of the", secondFieldName, "attribute.");
```

File: src/oop/object_independence/medium/q5/MainTest.java [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+        Object setMethodOutput = testClass.callMethod(setSecondField, setArguments, setMethodModifiers, classInstance);
+        _assertEquals(update, testClass.getFieldValue(classInstance, secondFieldName), incorrectSetMethodMessage);
+        assertNull(setMethodOutput, String.join(" ", "Your", setSecondField, "should not return any output"));
```
Comment (SamamthaH) [inline src/oop/object_independence/medium/q5/MainTest.java:121]:
```suggestion
        assertNull(setMethodOutput, String.join(" ", "Your", setSecondField, "method should not return any output"));
```

File: src/oop/object_independence/medium/q5/MainTest.java [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+        };
+        Object classInstance = testClass.createInstance(arguments);
+        _assertEquals(0, testClass.callMethod(getSecondField, classInstance),
+                "Your " + getSecondField + " method does not return the correct value.");
+        Object[][] setArguments = {
+                {value2, int.class}
+        };
+        testClass.callMethod(setSecondField, setArguments, classInstance);
+        _assertEquals(value2, testClass.callMethod(getSecondField, classInstance),
+                "Your " + getSecondField + " method does not return the correct value calling the " + setSecondField + " method.");
```
Comment (SamamthaH) [inline src/oop/object_independence/medium/q5/MainTest.java:155]:
```suggestion
                "Your " + getSecondField + " method does not return the correct value after calling the " + setSecondField + " method.");
```

File: src/oop/object_independence/medium/q5/MainTest.java [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+        String expected = "Team " + value1 + " has won " + value2 + " games";
+        String incorrectToStringMessage = String.join(" ",
+                "Your", className, " toString method does not return the correct String after updating the values of its attributes using its setter methods.");
```
Comment (SamamthaH) [inline src/oop/object_independence/medium/q5/MainTest.java:158]:
```suggestion
                "Your", className, " toString method does not return the correct String after updating the values of its attributes using its setter method.");
```

**Review by SamamthaH — CHANGES_REQUESTED** [review]

File: src/oop/object_independence/medium/q5/MainTest.java [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+    public void teamClassHasCorrectToStringMethod(String value1, int value2) throws Throwable {
+        Object[][] arguments = {
+                {value1, String.class},
+                {value2, int.class}
+        };
+        Object classInstance = testClass.createInstance(arguments);
+        String[] methodModifiers = {"public"};
+        String expected = "Team " + value1 + " has won " + value2 + " games";
+        String incorrectToStringMessage = String.join(" ",
+                "Your", className, " toString method does not return the correct String.");
```
Comment (SamamthaH) [inline src/oop/object_independence/medium/q5/MainTest.java:128]:
```suggestion
                "Your", className, "toString method does not return the correct String.");
```

File: src/oop/object_independence/medium/q5/MainTest.java [hunk 18]
```diff
@@ -0,0 +1,156 @@
...
+                "Your " + getSecondField + " method does not return the correct value.");
+        Object[][] setArguments = {
+                {value2, int.class}
+        };
+        testClass.callMethod(setSecondField, setArguments, classInstance);
+        _assertEquals(value2, testClass.callMethod(getSecondField, classInstance),
+                "Your " + getSecondField + " method does not return the correct value after calling the " + setSecondField + " method.");
+        String expected = "Team " + value1 + " has won " + value2 + " games";
+        String incorrectToStringMessage = String.join(" ",
+                "Your", className, " toString method does not return the correct String after updating the values of its attributes using its setter method.");
```
Comment (SamamthaH) [inline src/oop/object_independence/medium/q5/MainTest.java:151]:
```suggestion
                "Your", className, "toString method does not return the correct String after updating the values of its attributes using its setter method.");
```

**Review by SamamthaH — CHANGES_REQUESTED** [review]

File: src/oop/object_independence/medium/q5/MainTest.java [hunk 19]
```diff
@@ -0,0 +1,156 @@
...
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+import java.util.stream.Stream;
+
+import static global.tools.CustomAssertions._assertEquals;
+import static org.junit.jupiter.api.Assertions.*;
+
+public class MainTest {
+    // Java
```
[the remark this answers is not in the document]
Comment (SamamthaH) [reply src/oop/object_independence/medium/q5/MainTest.java:16]: In this comment https://github.com/canvas-gamification/canvas-gamification-junit-tests/pull/675#discussion_r1213821504 I asked you to fix the formatting on the parsons comment, however you switched it to Java instead. Is there another reason for the switch?

File: src/oop/object_independence/medium/q5/MainTest.java [hunk 20]
```diff
@@ -0,0 +1,156 @@
...
+    private final String setSecondField = "setTotalWins";
+
+    @BeforeEach
+    public void setUp() {
+        String classString = "oop.object_independence.medium.q5." + className;
+        testClass = new ObjectTest(classString);
+    }
+
+    @Test
+    public void teamClassHasRequiredFields() {
```
Comment (SamamthaH) [inline src/oop/object_independence/medium/q5/MainTest.java:31]:
```suggestion
    public void teamClassHasRequiredAttributes() {
```

File: src/oop/object_independence/medium/q5/MainTest.java [hunk 21]
```diff
@@ -0,0 +1,156 @@
...
+    private final String getSecondField = "getTotalWins";
+    private final String setSecondField = "setTotalWins";
+
+    @BeforeEach
+    public void setUp() {
+        String classString = "oop.object_independence.medium.q5." + className;
+        testClass = new ObjectTest(classString);
+    }
+
+    @Test
```
Comment (SamamthaH) [inline src/oop/object_independence/medium/q5/MainTest.java:30]: modify to include new hasField method

**Review by SamamthaH — CHANGES_REQUESTED** [review]

File: src/oop/object_independence/medium/q5/MainTest.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (KaiKhademi) [reply src/oop/object_independence/medium/q5/MainTest.java:16]: Same reason as the last prs

File: src/oop/object_independence/medium/q5/QuestionText.md [hunk 22]
```diff
@@ -0,0 +1,33 @@
+# oop-object_independence-medium-q5
+
+## Question Info
+
+- Question type: Java
+
+### Question Text
+
+Your friends are interested in betting on which team wins the Stanley Cup this year. Make a program called **Team** 
+where you keep track of the teams in the league and the games they have won as attribute.
```
Comment (Hedgemon4) [inline src/oop/object_independence/medium/q5/QuestionText.md:10]:
```suggestion
where you keep track of the teams in the league and the games they have won as attribute using the description below.
```

File: src/oop/object_independence/medium/q5/QuestionText.md [hunk 23]
```diff
@@ -0,0 +1,33 @@
...
+    - This constructor should be visible from another class
+
+- Methods:
+    - a `getTotalWins` method which is accessible from other classes and returns the **totalWins** attribute
+    - a `setTotalWins` method which accepts an `int` value, updates the **totalWins** attribute with it,
+      and is accessible from other classes
+    - a `toString` method which returns a String of the information about the `Team` class of the form
+      `"Team x has won y games"`, where x is the value of **name** and y is the value of **totalWins**,
+      and is visible from other classes.
+    - Note that all setter methods must not return any value
```
Comment (Hedgemon4) [inline src/oop/object_independence/medium/q5/QuestionText.md:33]: As there is not multiple setter methods, I would just add this point to the setter method description.

File: src/oop/object_independence/medium/q5/MainTest.java [hunk 24]
```diff
@@ -0,0 +1,197 @@
...
+        assertTrue(testClass.hasReturnType(getSecondField, null, int.class), String.format(incorrectReturnType, getSecondField, className));
+        Object classInstance = testClass.createInstance(arguments);
+        Object getMethodOutput = testClass.callMethod(getSecondField, getMethodModifiers, classInstance);
+        _assertEquals(value2, getMethodOutput, incorrectGetMethodMessage);
+    }
+
+    private static Stream<Arguments> set1InputProvider() {
+        return Stream.of(
+                Arguments.of(5, 6),
+                Arguments.of(10, 20),
```
Comment (Hedgemon4) [inline src/oop/object_independence/medium/q5/MainTest.java:112]: Should also test if you pass the same value that it already is, as well as setting the value to 0 and setting the value from 0 to another number.

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

**Review by Hedgemon4 — APPROVED** [review]

**Pull request merged**